src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -222,43 +227,143 @@ def _push_content_single_part(self) -> str:
         """
-        parameters = dict(object_key=f'manifests/{uuid4()}.tsv',
-                          data=self._construct_tsv_content().encode(),
-                          content_type='text/tab-separated-values')
+        if self.format == 'tsv':
+            data = self._construct_tsv_content().encode()
+            content_type = 'text/tab-separated-values'
+            object_key = f'manifests/{uuid4()}.tsv'
+        elif self.format == 'bdbag':
+            data = self._construct_bdbag()
+            content_type = 'application/zip'
+            object_key = f'manifests/{uuid4()}.zip'
+        else:
+            assert False
+        parameters = dict(object_key=object_key,
+                          data=data,
+                          content_type=content_type)
         return self.storage_service.put(**parameters)
 
     def _construct_tsv_content(self):
         es_search = self.es_search
 
         output = StringIO()
         writer = csv.writer(output, dialect='excel-tab')
-
         writer.writerow(list(self.manifest_entries['bundles'].keys()) +
                         list(self.manifest_entries['contents.files'].keys()))
         for hit in es_search.scan():
-            self._iterate_hit(hit, writer)
+            self._iterate_hit_tsv(hit, writer)
 
         return output.getvalue()
 
-    def _iterate_hit(self, es_search_hit, writer):
+    def _construct_bdbag(self) -> StringIO:
+        """
+        Create and return a file object of the sample data.
+        """
+        es_search = self.es_search
+        sample_file_object = StringIO()
+        sample_writer = csv.writer(sample_file_object, dialect='excel-tab')
+        # Add fieldnames.
+        sample_writer.writerow(list(chain(self.manifest_entries['contents.specimens'].keys(),
+                                          self.manifest_entries['contents.cell_suspensions'].keys(),
+                                          self.manifest_entries['bundles'].keys(),
+                                          self.manifest_entries['contents.files'].keys(),
+                                          ['file_url'])))
+        for hit in es_search.scan():
+            self._iterate_hit_bdbag(hit, sample_writer)
+
+        return self._create_zipped_bdbag(sample_file_object)
+
+    @staticmethod
+    def _create_zipped_bdbag(sample_file_object: StringIO) -> str:
+        """
+        Create participant data from sample data file object, write both to disk, and create and return BDBag archive.
+
+        :parameter sample_file_object: file objects of sample data
+        :returns bdbag: archived (zipped) BDBag with values of participant and sample data in payload
+        """
+
+        with TemporaryDirectory() as bag_path, TemporaryDirectory() as tsv_file_dir:
+            bag = bdbag_api.make_bag(bag_path)
+            data_path = os.path.join(bag_path, 'data')
+
+            # Get sample data from file object and from it create unique participant data.
```

Comment:
Still not right. I asked you explicitly NOT to parse the finished sample TSV. Maintain a set of participants as you populate `sample_file_object` and pass that set to this function.

Reply:
Changed it so that participant is now constructed in the same loop as sample, but as a set.